File: Programmers/Lv. 3/디스크 컨트롤러.py
import heapq

def solution(jobs):
    answer = 0

    workHeap = []

    jobs.sort(key=lambda x : x[0]) 

    start, now = -1, 0
    idx = 0
    while idx < len(jobs):
        for job in jobs:
            if start < job[0] <= now:
                heapq.heappush(workHeap, (job[1], job[0]))

        if workHeap:
            workTime, requestTime = heapq.heappop(workHeap)
            start = now
            now += workTime
            answer += now - requestTime
            idx += 1
        else:
            now += 1

    return answer // len(jobs)


import heapq
from collections import deque

def solution(jobs):
    answer = 0

    workHeap = []

    jobs.sort(key=lambda x : x[0]) 
    jobQueue = deque(jobs)
    
    start, now = -1, 0
    idx = 0
    while idx < len(jobs):
        while jobQueue:
            if start < jobQueue[0][0] <= now:
                requestTime, workTime = jobQueue.popleft()
                heapq.heappush(workHeap, (workTime, requestTime))
            else:
                break

        if workHeap:
            workTime, requestTime = heapq.heappop(workHeap)
            start = now
            now += workTime
            answer += now - requestTime
            idx += 1
        else:
            now += 1

    return answer // len(jobs)

File: Programmers/Lv. 3/디스크 컨트롤러.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

int solution(vector<vector<int>> jobs) {
    int answer = 0;
    
    queue<vector<int>> jobs_q;
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> jobs_pq;
    
    sort(jobs.begin(), jobs.end());
    for (auto job : jobs)
    {
        jobs_q.push(job);
    }
    
    int start = -1;
    int now = 0;
    int idx = 0;
    while (idx < jobs.size())
    {
        while (!jobs_q.empty())
        {
            if (start < jobs_q.front()[0] && jobs_q.front()[0] <= now)
            {
                jobs_pq.push({jobs_q.front()[1], jobs_q.front()[0]});
                jobs_q.pop();
            }
            else
            {
                break;
            }
        }

        if (!jobs_pq.empty())
        {
            int work_time = jobs_pq.top().first;
            int request_time = jobs_pq.top().second;            
            jobs_pq.pop();

            start = now;
            now += work_time;
            answer += (now - request_time);
            idx++;
        }
        else
        {
            now++;
        }
    }
    
    
    return int(answer / jobs.size());
}